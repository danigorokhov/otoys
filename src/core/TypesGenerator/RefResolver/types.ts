import {
    OpenAPIObject,
    PathsObject,
    ComponentsObject,
} from 'openapi3-ts/oas30';

export type RefParsed = {
    root: OpenAPIObject;
    path: string[];
};

export type PathItemObjectResolved = PathsObject[keyof PathsObject];

type ComponentsObjectRequired = Required<ComponentsObject>

export type RequestBodyObjectResolved = ComponentsObjectRequired['requestBodies'][
    keyof ComponentsObjectRequired['requestBodies']
];

export type ResponseObjectResolved = ComponentsObjectRequired['responses'][
    keyof ComponentsObjectRequired['responses']
];
