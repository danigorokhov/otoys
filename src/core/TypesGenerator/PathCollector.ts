import { PathsObject, PathItemObject } from 'openapi3-ts';

import { Config } from '../Config';

export class PathCollector {
    constructor(private paths: PathsObject, private whitelist: Config['pathWhitelist']) {}

    public collect() {
        const collectedPaths: PathItemObject[] = [];

        let whitelistRegExp: RegExp | null = null;
        if (this.whitelist) {
            whitelistRegExp = new RegExp(this.whitelist);
        }

        const pathEntries = Object.entries<PathItemObject>(this.paths);

        for (const [path, pathItem] of pathEntries) {
            if (this.matchPath(path, whitelistRegExp)) {
                collectedPaths.push(pathItem);
            }
        }

        return collectedPaths;
    }

    private matchPath(path: string, regExp: RegExp | null) {
        if (!regExp) return true;

        return Boolean(
            path.match(regExp),
        );
    }
}
