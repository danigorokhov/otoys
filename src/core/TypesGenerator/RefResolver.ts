import { URL } from 'url';
import {
    OpenAPIObject,
    ResponsesObject,
    RequestBodyObject,
    PathItemObject,
} from 'openapi3-ts/oas30';
import get from 'lodash/get';

import {
    isPathItemObject,
    isResponsesObject,
    isRequestBodyObject,
    isSchemaObject,
} from '../utils/typePredicates';
import { SchemaEntry } from '../types/schema';

type RefParsed = {
    root: OpenAPIObject;
    path: string[];
};

// TODO maybe on higher level along with entire schema validation
type ValidateSchemaName = (name: string) => void;

const validateSchemaName: ValidateSchemaName = name => {
    if (name.match(/\W/)) {
        throw new Error(
            'validateSchemaName: schema name must consists of ASCII based characters'
        );
    }
}

export class RefResolver {
    constructor(private document: OpenAPIObject) {}

    public isValidURL(ref: string): boolean {
        try {
            new URL(ref);
        } catch (error) {
            // TODO https://stackoverflow.com/q/75382275/13215082
            if (error !== null && typeof error === 'object' && 'code' in error && error.code === 'ERR_INVALID_URL') {
                return false;
            }

            throw error;
        }

        return true;
    }

    /**
     * Validate reference
     * @returns reference in format `#[\w/]+`
     */
    public validateRef(ref: string): string {
        if (this.isValidURL(ref)) {
            throw new Error(
                `RefResolver: reference to schema by URL isn't supported. Passed reference: ${ref}`
            );
        }

        if (!ref.startsWith('#')) {
            throw new Error(
                `RefResolver: reference to schema by different file isn't supported. Passed reference: ${ref}`
            );
        }

        if (!ref.match(/^#[\w/]+$/)) {
            throw new Error(
                `RefResolver: reference has invalid format. Passed reference: ${ref}`
            );
        }

        return ref;
    }

    public parseRef(ref: string): RefParsed {
        const refValidate = this.validateRef(ref);

        const pathRaw = refValidate
            // cut '#', since validateRef returns local references
            .substring(1)
            .split('/');

        const path = pathRaw.filter(pathPart => pathPart.length !== 0);

        return {
            root: this.document,
            path,
        };
    }

    // TODO support refs to document.paths['/path/with/slashes/inside']
    public resolvePath(ref: string): PathItemObject {
        const {
            root,
            path: refPath,
        } = this.parseRef(ref);

        const path: unknown = get(root, refPath, null);

        if (path === null) {
            throw new Error(
                `RefResolver: path resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isPathItemObject(path)) {
            throw new Error(
                `RefResolver: path resolving failed. Resolved object isn't of type PathItemObject. Passed reference: ${ref}`
            );
        }

        return path;
    }

    public resolveRequestBody(ref: string): RequestBodyObject {
        const {
            root,
            path,
        } = this.parseRef(ref);

        const requestBody: unknown = get(root, path, null);

        if (requestBody === null) {
            throw new Error(
                `RefResolver: requestBody resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isRequestBodyObject(requestBody)) {
            throw new Error(
                `RefResolver: requestBody resolving failed. Resolved object isn't of type RequestBodyObject. Passed reference: ${ref}`
            );
        }

        return requestBody;
    }

    public resolveResponses(ref: string): ResponsesObject {
        const {
            root,
            path,
        } = this.parseRef(ref);

        const responses: unknown = get(root, path, null);

        if (responses === null) {
            throw new Error(
                `RefResolver: responses resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isResponsesObject(responses)) {
            throw new Error(
                `RefResolver: responses resolving failed. Resolved object isn't of type ResponsesObject. Passed reference: ${ref}`
            );
        }

        return responses;
    }

    public resolveSchema(ref: string): SchemaEntry {
        const {
            root,
            path,
        } = this.parseRef(ref);

        const schemaObject: unknown = get(root, path, null);

        if (schemaObject === null) {
            throw new Error(
                `RefResolver: schema resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isSchemaObject(schemaObject)) {
            throw new Error(
                `RefResolver: schema resolving failed. Resolved object isn't of type SchemaObject. Passed reference: ${ref}`
            );
        }

        const schemaName = path.at(-1) ?? null;

        if (schemaName === null) {
            throw new Error(
                `RefResolver: schema resolving failed. Schema name isn't valid. Passed reference: ${ref}`
            );
        }
        validateSchemaName(schemaName);

        return [schemaName, schemaObject];
    }
}
