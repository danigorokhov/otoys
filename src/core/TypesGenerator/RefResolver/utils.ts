import { SchemaName } from '../../types/schema';

// TODO maybe on higher level along with entire schema validation
type ValidateSchemaName = (name: SchemaName) => void;

export const validateSchemaName: ValidateSchemaName = name => {
    if (name.match(/\W/)) {
        throw new Error(
            'validateSchemaName: schema name must consists of ASCII based characters'
        );
    }
}
