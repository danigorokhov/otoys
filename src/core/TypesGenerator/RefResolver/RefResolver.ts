import { URL } from 'url';
import {
    OpenAPIObject,
    ResponseObject,
    RequestBodyObject,
    isReferenceObject,
} from 'openapi3-ts/oas30';
import get from 'lodash/get';

import {
    isPathItemObject,
    isResponseObject,
    isRequestBodyObject,
    isSchemaObject,
} from '../../utils/typePredicates';
import { SchemaEntry } from '../../types/schema';
import {
    RefParsed,
    PathItemObjectResolved,
    RequestBodyObjectResolved,
    ResponseObjectResolved,
} from './types';
import { validateSchemaName } from './utils';

/**
 * Resolves objects in document by reference. Reference corresponds RFC3986.
 */
export class RefResolver {
    constructor(private document: OpenAPIObject) {}

    public isValidURL(ref: string): boolean {
        try {
            new URL(ref);
        } catch (error) {
            // TODO https://stackoverflow.com/q/75382275/13215082
            if (error !== null && typeof error === 'object' && 'code' in error && error.code === 'ERR_INVALID_URL') {
                return false;
            }

            throw error;
        }

        return true;
    }

    /**
     * Validates reference
     * @returns reference in format `#[\w/]+`
     */
    public validateRef(ref: string): string {
        if (this.isValidURL(ref)) {
            throw new Error(
                `RefResolver: reference to schema by URL isn't supported. Passed reference: ${ref}`
            );
        }

        if (!ref.startsWith('#')) {
            throw new Error(
                `RefResolver: reference to schema by different file isn't supported. Passed reference: ${ref}`
            );
        }

        if (!ref.match(/^#[\w/]+$/)) {
            throw new Error(
                `RefResolver: reference has invalid format. Passed reference: ${ref}`
            );
        }

        return ref;
    }

    public parseRef(ref: string): RefParsed {
        const refValidate = this.validateRef(ref);

        const pathRaw = refValidate
            // cut '#', since validateRef returns local references
            .substring(1)
            .split('/');

        // https://swagger.io/docs/specification/using-ref/#escape
        const pathUnescaped = pathRaw.map(pathPart => {
            // Order is worth. If tilde is unescaped firstly, it can be created one more '~1' combination.
            // For instance: '/test~01' becomes '/test/' instead of '/test~1'.
            const withUnescapedSlash = pathPart.replaceAll('~1', '/');
            const withUnescapedTilde = withUnescapedSlash.replaceAll('~0', '~');

            return withUnescapedTilde;
        });

        const path = pathUnescaped.filter(pathPart => pathPart.length !== 0);

        return {
            root: this.document,
            path,
        };
    }

    public resolvePath(ref: string): PathItemObjectResolved {
        const {
            root,
            path: refPath,
        } = this.parseRef(ref);

        const path: unknown = get(root, refPath, null);

        if (path === null) {
            throw new Error(
                `RefResolver: path resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isPathItemObject(path)) {
            throw new Error(
                `RefResolver: path resolving failed. Resolved object isn't of type PathItemObject. Passed reference: ${ref}`
            );
        }

        return path;
    }

    public resolveRequestBody(ref: string): RequestBodyObjectResolved {
        const {
            root,
            path,
        } = this.parseRef(ref);

        const requestBody: unknown = get(root, path, null);

        if (requestBody === null) {
            throw new Error(
                `RefResolver: requestBody resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isRequestBodyObject(requestBody) && !isReferenceObject(requestBody)) {
            throw new Error(
                `RefResolver: requestBody resolving failed. Resolved object isn't of type RequestBodyObject or ReferenceObject. Passed reference: ${ref}`
            );
        }

        return requestBody;
    }

    public resolveRequestBodyDeep(ref: string, limit: number = 100): RequestBodyObject {
        let counter = 0;
        let requestBody = this.resolveRequestBody(ref);

        while (counter < limit && isReferenceObject(requestBody)) {
            requestBody = this.resolveRequestBody(requestBody.$ref);

            counter++;
        }

        if (counter >= limit) {
            throw new Error(
                `RefResolver: requestBody resolving failed. Too many nested references, limit ${limit} exceeded. Passed reference: ${ref}`,
            );
        }

        if (isReferenceObject(requestBody)) {
            throw new Error(
                `RefResolver: unexpected error, requestBody is reference. Passed reference: ${ref}`,
            );
        }

        return requestBody;
    }

    public resolveResponse(ref: string): ResponseObjectResolved {
        const {
            root,
            path,
        } = this.parseRef(ref);

        const response: unknown = get(root, path, null);

        if (response === null) {
            throw new Error(
                `RefResolver: response resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isResponseObject(response) && !isReferenceObject(response)) {
            throw new Error(
                `RefResolver: response resolving failed. Resolved object isn't of type ResponseObject or ReferenceObject. Passed reference: ${ref}`
            );
        }

        return response;
    }

    public resolveResponseDeep(ref: string, limit: number = 100): ResponseObject {
        let counter = 0;
        let responseObject = this.resolveResponse(ref);

        while (counter < limit && isReferenceObject(responseObject)) {
            responseObject = this.resolveResponse(responseObject.$ref);

            counter++;
        }

        if (counter >= limit) {
            throw new Error(
                `RefResolver: response resolving failed. Too many nested references, limit ${limit} exceeded. Passed reference: ${ref}`,
            );
        }

        if (isReferenceObject(responseObject)) {
            throw new Error(
                `RefResolver: unexpected error, response is reference. Passed reference: ${ref}`,
            );
        }

        return responseObject;
    }

    public resolveSchema(ref: string): SchemaEntry {
        const {
            root,
            path,
        } = this.parseRef(ref);

        const schemaObject: unknown = get(root, path, null);

        if (schemaObject === null) {
            throw new Error(
                `RefResolver: schema resolving failed. Cannot resolve object by passed reference. Passed reference: ${ref}`
            );
        }

        if (!isSchemaObject(schemaObject)) {
            throw new Error(
                `RefResolver: schema resolving failed. Resolved object isn't of type SchemaObject. Passed reference: ${ref}`
            );
        }

        const schemaName = path.at(-1) ?? null;

        if (schemaName === null) {
            throw new Error(
                `RefResolver: schema resolving failed. Schema name isn't valid. Passed reference: ${ref}`
            );
        }
        validateSchemaName(schemaName);

        return [schemaName, schemaObject];
    }
}
