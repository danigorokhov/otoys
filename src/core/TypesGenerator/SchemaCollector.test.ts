import { expect, describe, it } from '@jest/globals';
import { OpenAPIObject } from 'openapi3-ts/oas30';

import { SchemaCollector } from './SchemaCollector';
import { RefResolver } from './RefResolver';
import { PathCollector } from './PathCollector';

import petstore3 from '../../../examples/v3/petstore3.json';

describe('TypesGenerator/TypesGeneratorV3/SchemaCollector', () => {
    // TODO more detailed tests
    it('should collect schemas', () => {
        // TODO figure out why petstore3 cannot be assigned to OpenAPIObject
        const document = petstore3 as unknown as OpenAPIObject;
        const resolver = new RefResolver(document);
        const pathCollector = new PathCollector(document.paths, undefined);
        const paths = pathCollector.collect();
        const schemaCollector = new SchemaCollector(paths, resolver);

        expect(schemaCollector.collect()).toMatchSnapshot();
    });
});
