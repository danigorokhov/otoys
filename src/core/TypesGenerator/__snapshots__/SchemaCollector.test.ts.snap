// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypesGenerator/TypesGeneratorV3/SchemaCollector should collect schemas for one-password-events API 1`] = `
Map {
  "UnnamedSchema1" => {
    "oneOf": [
      {
        "$ref": "#/components/schemas/Cursor",
      },
      {
        "$ref": "#/components/schemas/ResetCursor",
      },
    ],
  },
  "UnnamedSchema2" => {
    "oneOf": [
      {
        "$ref": "#/components/schemas/Cursor",
      },
      {
        "$ref": "#/components/schemas/ResetCursor",
      },
    ],
  },
  "Introspection" => {
    "properties": {
      "Features": {
        "example": [
          "itemusages",
          "signinattempts",
        ],
        "items": {
          "type": "string",
        },
        "type": "array",
      },
      "IssuedAt": {
        "$ref": "#/components/schemas/DateTimeRFC3339",
      },
      "UUID": {
        "type": "string",
      },
    },
    "type": "object",
  },
  "Error" => {
    "properties": {
      "Error": {
        "properties": {
          "Message": {
            "description": "The error message.",
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "type": "object",
  },
  "ItemUsageItems" => {
    "allOf": [
      {
        "properties": {
          "items": {
            "items": {
              "$ref": "#/components/schemas/ItemUsage",
            },
            "type": "array",
          },
        },
      },
      {
        "$ref": "#/components/schemas/CursorCollection",
      },
    ],
    "description": "An object wrapping cursor properties and a list of items usages",
  },
  "SignInAttemptItems" => {
    "allOf": [
      {
        "properties": {
          "items": {
            "items": {
              "$ref": "#/components/schemas/SignInAttempt",
            },
            "type": "array",
          },
        },
      },
      {
        "$ref": "#/components/schemas/CursorCollection",
      },
    ],
    "description": "An object wrapping cursor properties and a list of sign-in attempts",
  },
  "Cursor" => {
    "description": "Cursor",
    "properties": {
      "cursor": {
        "description": "Cursor to fetch more data if available or continue the polling process if required",
        "example": "aGVsbG8hIGlzIGl0IG1lIHlvdSBhcmUgbG9va2luZyBmb3IK",
        "type": "string",
      },
    },
  },
  "ResetCursor" => {
    "description": "Reset cursor",
    "properties": {
      "end_time": {
        "$ref": "#/components/schemas/DateTimeRFC3339",
      },
      "limit": {
        "maximum": 1000,
        "minimum": 1,
        "type": "number",
      },
      "start_time": {
        "$ref": "#/components/schemas/DateTimeRFC3339",
      },
    },
  },
  "AuxillarySchema1" => {
    "example": [
      "itemusages",
      "signinattempts",
    ],
    "items": {
      "type": "string",
    },
    "type": "array",
  },
  "DateTimeRFC3339" => {
    "example": "2020-06-11T16:32:50-03:00",
    "format": "date-time",
    "type": "string",
  },
  "AuxillarySchema2" => {
    "type": "string",
  },
  "AuxillarySchema3" => {
    "properties": {
      "Message": {
        "description": "The error message.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "AuxillarySchema4" => {
    "properties": {
      "items": {
        "items": {
          "$ref": "#/components/schemas/ItemUsage",
        },
        "type": "array",
      },
    },
  },
  "CursorCollection" => {
    "allOf": [
      {
        "$ref": "#/components/schemas/Cursor",
      },
      {
        "properties": {
          "has_more": {
            "description": "Whether there may still be more data to fetch using the returned cursor. If true, the subsequent request could still be empty.",
            "type": "boolean",
          },
        },
      },
    ],
    "description": "Common cursor properties for collection responses",
  },
  "AuxillarySchema5" => {
    "properties": {
      "items": {
        "items": {
          "$ref": "#/components/schemas/SignInAttempt",
        },
        "type": "array",
      },
    },
  },
  "AuxillarySchema6" => {
    "description": "Cursor to fetch more data if available or continue the polling process if required",
    "example": "aGVsbG8hIGlzIGl0IG1lIHlvdSBhcmUgbG9va2luZyBmb3IK",
    "type": "string",
  },
  "AuxillarySchema7" => {
    "maximum": 1000,
    "minimum": 1,
    "type": "number",
  },
  "AuxillarySchema8" => {
    "type": "string",
  },
  "AuxillarySchema9" => {
    "description": "The error message.",
    "type": "string",
  },
  "AuxillarySchema10" => {
    "items": {
      "$ref": "#/components/schemas/ItemUsage",
    },
    "type": "array",
  },
  "AuxillarySchema11" => {
    "properties": {
      "has_more": {
        "description": "Whether there may still be more data to fetch using the returned cursor. If true, the subsequent request could still be empty.",
        "type": "boolean",
      },
    },
  },
  "AuxillarySchema12" => {
    "items": {
      "$ref": "#/components/schemas/SignInAttempt",
    },
    "type": "array",
  },
  "ItemUsage" => {
    "description": "A single item usage object",
    "properties": {
      "action": {
        "$ref": "#/components/schemas/Action",
      },
      "client": {
        "$ref": "#/components/schemas/Client",
      },
      "item_uuid": {
        "$ref": "#/components/schemas/UUID",
      },
      "timestamp": {
        "$ref": "#/components/schemas/DateTimeRFC3339",
      },
      "used_version": {
        "type": "integer",
      },
      "user": {
        "$ref": "#/components/schemas/User",
      },
      "uuid": {
        "$ref": "#/components/schemas/UUID",
      },
      "vault_uuid": {
        "$ref": "#/components/schemas/UUID",
      },
    },
  },
  "AuxillarySchema13" => {
    "description": "Whether there may still be more data to fetch using the returned cursor. If true, the subsequent request could still be empty.",
    "type": "boolean",
  },
  "SignInAttempt" => {
    "description": "A single sign-in attempt object",
    "properties": {
      "category": {
        "enum": [
          "success",
          "credentials_failed",
          "mfa_failed",
          "modern_version_failed",
          "firewall_failed",
          "firewall_reported_success",
        ],
        "example": "firewall_failed",
        "type": "string",
      },
      "client": {
        "$ref": "#/components/schemas/Client",
      },
      "country": {
        "description": "Country ISO Code",
        "example": "France",
        "type": "string",
      },
      "details": {
        "$ref": "#/components/schemas/Details",
      },
      "session_uuid": {
        "$ref": "#/components/schemas/UUID",
      },
      "target_user": {
        "$ref": "#/components/schemas/User",
      },
      "timestamp": {
        "$ref": "#/components/schemas/DateTimeRFC3339",
      },
      "type": {
        "enum": [
          "credentials_ok",
          "mfa_ok",
          "password_secret_bad",
          "mfa_missing",
          "totp_disabled",
          "totp_bad",
          "totp_timeout",
          "u2f_disabled",
          "u2f_bad",
          "u2f_timout",
          "duo_disabled",
          "duo_bad",
          "duo_timeout",
          "duo_native_bad",
          "platform_secret_disabled",
          "platform_secret_bad",
          "platform_secret_proxy",
          "code_disabled",
          "code_bad",
          "code_timeout",
          "ip_blocked",
          "continent_blocked",
          "country_blocked",
          "anonymous_blocked",
          "all_blocked",
          "modern_version_missing",
          "modern_version_old",
        ],
        "example": "continent_blocked",
        "type": "string",
      },
      "uuid": {
        "$ref": "#/components/schemas/UUID",
      },
    },
  },
  "Action" => {},
  "Client" => {
    "description": "Metadata gathered about the client",
    "properties": {
      "app_name": {
        "example": "1Password Extension",
        "type": "string",
      },
      "app_version": {
        "example": "20127",
        "type": "string",
      },
      "ip_address": {
        "example": "13.227.95.22",
        "type": "string",
      },
      "os_name": {
        "example": "MacOSX",
        "type": "string",
      },
      "os_version": {
        "example": "10.15.6",
        "type": "string",
      },
      "platform_name": {
        "example": "Chrome",
        "type": "string",
      },
      "platform_version": {
        "description": "Depending on the platform used, this can be the version of the browser that the client extension is installed, the model of computer that the native application is installed or the machine's CPU version that the CLI was installed",
        "type": "string",
      },
    },
  },
  "UUID" => {
    "example": "56YE2TYN2VFYRLNSHKPW5NVT5E",
    "type": "string",
  },
  "AuxillarySchema14" => {
    "type": "integer",
  },
  "User" => {
    "description": "User object",
    "properties": {
      "email": {
        "format": "email",
        "type": "string",
      },
      "name": {
        "description": "Full name",
        "example": "Jack O'Neill",
        "type": "string",
      },
      "uuid": {
        "$ref": "#/components/schemas/UUID",
      },
    },
  },
  "AuxillarySchema15" => {
    "enum": [
      "success",
      "credentials_failed",
      "mfa_failed",
      "modern_version_failed",
      "firewall_failed",
      "firewall_reported_success",
    ],
    "example": "firewall_failed",
    "type": "string",
  },
  "AuxillarySchema16" => {
    "description": "Country ISO Code",
    "example": "France",
    "type": "string",
  },
  "Details" => {
    "description": "Additional information about the sign-in attempt",
    "properties": {
      "value": {
        "description": "For firewall prevented sign-ins, the value is the chosen continent, country, etc. that blocked the sign-in attempt",
        "example": "Europe",
        "type": "string",
      },
    },
  },
  "AuxillarySchema17" => {
    "enum": [
      "credentials_ok",
      "mfa_ok",
      "password_secret_bad",
      "mfa_missing",
      "totp_disabled",
      "totp_bad",
      "totp_timeout",
      "u2f_disabled",
      "u2f_bad",
      "u2f_timout",
      "duo_disabled",
      "duo_bad",
      "duo_timeout",
      "duo_native_bad",
      "platform_secret_disabled",
      "platform_secret_bad",
      "platform_secret_proxy",
      "code_disabled",
      "code_bad",
      "code_timeout",
      "ip_blocked",
      "continent_blocked",
      "country_blocked",
      "anonymous_blocked",
      "all_blocked",
      "modern_version_missing",
      "modern_version_old",
    ],
    "example": "continent_blocked",
    "type": "string",
  },
  "AuxillarySchema18" => {
    "example": "1Password Extension",
    "type": "string",
  },
  "AuxillarySchema19" => {
    "example": "20127",
    "type": "string",
  },
  "AuxillarySchema20" => {
    "example": "13.227.95.22",
    "type": "string",
  },
  "AuxillarySchema21" => {
    "example": "MacOSX",
    "type": "string",
  },
  "AuxillarySchema22" => {
    "example": "10.15.6",
    "type": "string",
  },
  "AuxillarySchema23" => {
    "example": "Chrome",
    "type": "string",
  },
  "AuxillarySchema24" => {
    "description": "Depending on the platform used, this can be the version of the browser that the client extension is installed, the model of computer that the native application is installed or the machine's CPU version that the CLI was installed",
    "type": "string",
  },
  "AuxillarySchema25" => {
    "format": "email",
    "type": "string",
  },
  "AuxillarySchema26" => {
    "description": "Full name",
    "example": "Jack O'Neill",
    "type": "string",
  },
  "AuxillarySchema27" => {
    "description": "For firewall prevented sign-ins, the value is the chosen continent, country, etc. that blocked the sign-in attempt",
    "example": "Europe",
    "type": "string",
  },
}
`;

exports[`TypesGenerator/TypesGeneratorV3/SchemaCollector should collect schemas for petstore3 API 1`] = `
Map {
  "Pet" => {
    "properties": {
      "category": {
        "$ref": "#/components/schemas/Category",
      },
      "id": {
        "example": 10,
        "format": "int64",
        "type": "integer",
      },
      "name": {
        "example": "doggie",
        "type": "string",
      },
      "photoUrls": {
        "items": {
          "type": "string",
          "xml": {
            "name": "photoUrl",
          },
        },
        "type": "array",
        "xml": {
          "wrapped": true,
        },
      },
      "status": {
        "description": "pet status in the store",
        "enum": [
          "available",
          "pending",
          "sold",
        ],
        "type": "string",
      },
      "tags": {
        "items": {
          "$ref": "#/components/schemas/Tag",
        },
        "type": "array",
        "xml": {
          "wrapped": true,
        },
      },
    },
    "required": [
      "name",
      "photoUrls",
    ],
    "type": "object",
    "xml": {
      "name": "pet",
    },
  },
  "UnnamedSchema1" => {
    "format": "binary",
    "type": "string",
  },
  "Order" => {
    "properties": {
      "complete": {
        "type": "boolean",
      },
      "id": {
        "example": 10,
        "format": "int64",
        "type": "integer",
      },
      "petId": {
        "example": 198772,
        "format": "int64",
        "type": "integer",
      },
      "quantity": {
        "example": 7,
        "format": "int32",
        "type": "integer",
      },
      "shipDate": {
        "format": "date-time",
        "type": "string",
      },
      "status": {
        "description": "Order Status",
        "enum": [
          "placed",
          "approved",
          "delivered",
        ],
        "example": "approved",
        "type": "string",
      },
    },
    "type": "object",
    "xml": {
      "name": "order",
    },
  },
  "User" => {
    "properties": {
      "email": {
        "example": "john@email.com",
        "type": "string",
      },
      "firstName": {
        "example": "John",
        "type": "string",
      },
      "id": {
        "example": 10,
        "format": "int64",
        "type": "integer",
      },
      "lastName": {
        "example": "James",
        "type": "string",
      },
      "password": {
        "example": "12345",
        "type": "string",
      },
      "phone": {
        "example": "12345",
        "type": "string",
      },
      "userStatus": {
        "description": "User Status",
        "example": 1,
        "format": "int32",
        "type": "integer",
      },
      "username": {
        "example": "theUser",
        "type": "string",
      },
    },
    "type": "object",
    "xml": {
      "name": "user",
    },
  },
  "UnnamedSchema2" => {
    "items": {
      "$ref": "#/components/schemas/User",
    },
    "type": "array",
  },
  "UnnamedSchema3" => {
    "items": {
      "$ref": "#/components/schemas/Pet",
    },
    "type": "array",
  },
  "UnnamedSchema4" => {
    "items": {
      "$ref": "#/components/schemas/Pet",
    },
    "type": "array",
  },
  "UnnamedSchema5" => {
    "items": {
      "$ref": "#/components/schemas/Pet",
    },
    "type": "array",
  },
  "UnnamedSchema6" => {
    "items": {
      "$ref": "#/components/schemas/Pet",
    },
    "type": "array",
  },
  "ApiResponse" => {
    "properties": {
      "code": {
        "format": "int32",
        "type": "integer",
      },
      "message": {
        "type": "string",
      },
      "type": {
        "type": "string",
      },
    },
    "type": "object",
    "xml": {
      "name": "##default",
    },
  },
  "UnnamedSchema7" => {
    "additionalProperties": {
      "format": "int32",
      "type": "integer",
    },
    "type": "object",
  },
  "UnnamedSchema8" => {
    "type": "string",
  },
  "UnnamedSchema9" => {
    "type": "string",
  },
  "AuxillarySchema1" => {
    "example": 10,
    "format": "int64",
    "type": "integer",
  },
  "AuxillarySchema2" => {
    "example": "doggie",
    "type": "string",
  },
  "Category" => {
    "properties": {
      "id": {
        "example": 1,
        "format": "int64",
        "type": "integer",
      },
      "name": {
        "example": "Dogs",
        "type": "string",
      },
    },
    "type": "object",
    "xml": {
      "name": "category",
    },
  },
  "AuxillarySchema3" => {
    "items": {
      "type": "string",
      "xml": {
        "name": "photoUrl",
      },
    },
    "type": "array",
    "xml": {
      "wrapped": true,
    },
  },
  "AuxillarySchema4" => {
    "items": {
      "$ref": "#/components/schemas/Tag",
    },
    "type": "array",
    "xml": {
      "wrapped": true,
    },
  },
  "AuxillarySchema5" => {
    "description": "pet status in the store",
    "enum": [
      "available",
      "pending",
      "sold",
    ],
    "type": "string",
  },
  "AuxillarySchema6" => {
    "example": 10,
    "format": "int64",
    "type": "integer",
  },
  "AuxillarySchema7" => {
    "example": 198772,
    "format": "int64",
    "type": "integer",
  },
  "AuxillarySchema8" => {
    "example": 7,
    "format": "int32",
    "type": "integer",
  },
  "AuxillarySchema9" => {
    "format": "date-time",
    "type": "string",
  },
  "AuxillarySchema10" => {
    "description": "Order Status",
    "enum": [
      "placed",
      "approved",
      "delivered",
    ],
    "example": "approved",
    "type": "string",
  },
  "AuxillarySchema11" => {
    "type": "boolean",
  },
  "AuxillarySchema12" => {
    "example": 10,
    "format": "int64",
    "type": "integer",
  },
  "AuxillarySchema13" => {
    "example": "theUser",
    "type": "string",
  },
  "AuxillarySchema14" => {
    "example": "John",
    "type": "string",
  },
  "AuxillarySchema15" => {
    "example": "James",
    "type": "string",
  },
  "AuxillarySchema16" => {
    "example": "john@email.com",
    "type": "string",
  },
  "AuxillarySchema17" => {
    "example": "12345",
    "type": "string",
  },
  "AuxillarySchema18" => {
    "example": "12345",
    "type": "string",
  },
  "AuxillarySchema19" => {
    "description": "User Status",
    "example": 1,
    "format": "int32",
    "type": "integer",
  },
  "AuxillarySchema20" => {
    "format": "int32",
    "type": "integer",
  },
  "AuxillarySchema21" => {
    "type": "string",
  },
  "AuxillarySchema22" => {
    "type": "string",
  },
  "AuxillarySchema23" => {
    "format": "int32",
    "type": "integer",
  },
  "AuxillarySchema24" => {
    "example": 1,
    "format": "int64",
    "type": "integer",
  },
  "AuxillarySchema25" => {
    "example": "Dogs",
    "type": "string",
  },
  "AuxillarySchema26" => {
    "type": "string",
    "xml": {
      "name": "photoUrl",
    },
  },
  "Tag" => {
    "properties": {
      "id": {
        "format": "int64",
        "type": "integer",
      },
      "name": {
        "type": "string",
      },
    },
    "type": "object",
    "xml": {
      "name": "tag",
    },
  },
  "AuxillarySchema27" => {
    "format": "int64",
    "type": "integer",
  },
  "AuxillarySchema28" => {
    "type": "string",
  },
}
`;
